python/descriptor_pool.c

```diff
@@ -110,30 +121,63 @@ static PyObject* PyUpb_DescriptorPool_New(PyTypeObject* type, PyObject* args,
   return PyUpb_DescriptorPool_DoCreate(type, db);
 }
 
-/*
- * PyUpb_DescriptorPool_AddSerializedFile()
- *
- * Implements:
- *   DescriptorPool.AddSerializedFile(self, serialized_file_descriptor)
- *
- * Adds the given serialized FileDescriptorProto to the pool.
- */
-static PyObject* PyUpb_DescriptorPool_AddSerializedFile(
-    PyObject* _self, PyObject* serialized_pb) {
-  PyUpb_DescriptorPool* self = (PyUpb_DescriptorPool*)_self;
-  char* buf;
-  Py_ssize_t size;
-  upb_arena* arena = upb_arena_new();
-  if (!arena) PYUPB_RETURN_OOM;
-  PyObject* result = NULL;
+static PyObject* PyUpb_DescriptorPool_DoAdd(PyObject* _self,
+                                            PyObject* file_desc);
 
+static bool PyUpb_DescriptorPool_TryLoadFileProto(PyUpb_DescriptorPool* self,
+                                                  PyObject* proto) {
+  if (proto == NULL) {
+    if (PyErr_ExceptionMatches(PyExc_KeyError)) {
+      // Expected error: item was simply not found.
+      PyErr_Clear();
+      return true;  // We didn't accomplish our goal, but we didn't error out.
+    }
+    return false;
+  }
+  if (proto == Py_None) return true;
+  return PyUpb_DescriptorPool_DoAdd((PyObject*)self, proto) != NULL;
+}
+
+static bool PyUpb_DescriptorPool_TryLoadSymbol(PyUpb_DescriptorPool* self,
+                                               PyObject* sym) {
+  if (!self->db) return false;
+  PyObject* file_proto =
+      PyObject_CallMethod(self->db, "FindFileContainingSymbol", "O", sym);
+  bool ret = PyUpb_DescriptorPool_TryLoadFileProto(self, file_proto);
+  Py_XDECREF(file_proto);
+  return ret;
+}
+
+static bool PyUpb_DescriptorPool_TryLoadFilename(PyUpb_DescriptorPool* self,
+                                                 PyObject* filename) {
+  if (!self->db) return false;
+  PyObject* file_proto =
+      PyObject_CallMethod(self->db, "FindFileByName", "O", filename);
+  bool ret = PyUpb_DescriptorPool_TryLoadFileProto(self, file_proto);
+  Py_XDECREF(file_proto);
+  return ret;
+}
+
+bool PyUpb_DescriptorPool_CheckNoDatabase(PyObject* _self) {
```

Comment:
CheckNoDatabase has too many negatives to easily scan and "Check" is confusable with CHECK failing inside google.

Let's rename this 

```
bool PyUpb_DescriptorPool_HasDatabase(...) {
  return self->db != null;
}
```

Having now written that down, I begin to think maybe we wouldn't have a helper method here.  Maybe the 2 spots that call this should inline their test more explicitly and even provide more specialized error messages.

Reply:
Done.